pkg/export/setup/setup_test.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package setup
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/GoogleCloudPlatform/prometheus-engine/pkg/export"
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/go-kit/log"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestFromFlags_NotOnGCE(t *testing.T) {
+	// Asserting there is actually no GCE underneath.
+	if metadata.OnGCE() {
+		t.Fatal("This test assumes we don't run on GCP")
+	}
+
+	fake := kingpin.New("test", "test")
+	newExport := FromFlags(fake, "product")
+	// Our readines is default (3 * 10s), so ensure FromFlags is never longer than 30s.
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	// Fake app invocation.
+	if _, err := fake.Parse(nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// Make sure constructor does is not stuck forever.
+	_, _ = newExport(ctx, log.NewLogfmtLogger(os.Stderr), nil)
+}
+
+// Regression test for b/344740239. We ensure that even stuck metadata servers
+// calls will timeout correctly (we propagate context properly).
+func TestTryPopulateUnspecifiedFromMetadata(t *testing.T) {
+	// Asserting there is actually no GCE underneath.
+	if metadata.OnGCE() {
```

Comment:
Suggestion: What if we just mock the metadata server? e.g. we create an interface and pass that to `tryPopulateUnspecifiedFromMetadata`?

Reply:
Yea, what's the advantage of this versus what's now?

We can mock, but then we just test if we pass context correctly.

 Here we test the metadata client to also pass context correctly. I know we shouldn't generally test not our code.. but well, our product depends on this we just got regression because of that, thus let's maybe test this case?